packages/query/src/index.ts

```diff
@@ -114,73 +120,160 @@ export class LiveQuery extends TxProcessor implements Client {
     return await this.client.findOne(_class, query, options)
   }
 
-  query<T extends Doc>(
+  private findQuery<T extends Doc>(
+    _class: Ref<Class<T>>,
+    query: DocumentQuery<T>,
+    options?: FindOptions<T>
+  ): Query | undefined {
+    const queries = this.queries.get(_class)
+    if (queries === undefined) return
+    for (const q of queries) {
+      if (!deepEqual(query, q.query) || !deepEqual(options, q.options)) continue
+      return q
+    }
+  }
+
+  private removeFromQueue (q: Query): void {
+    if (q.callbacks.length === 0) {
+      const queueIndex = this.queue.indexOf(q)
+      if (queueIndex !== -1) {
+        this.queue.splice(queueIndex, 1)
+      }
+    }
+  }
+
+  private pushCallback (q: Query, callback: (result: Doc[]) => void): void {
+    q.callbacks.push(callback)
+    setTimeout(async () => {
+      if (q !== undefined) {
+        await this.callback(q)
+      }
+    }, 0)
+  }
+
+  private getQuery<T extends Doc>(
+    _class: Ref<Class<T>>,
+    query: DocumentQuery<T>,
+    callback: (result: Doc[]) => void,
+    options?: FindOptions<T>
+  ): Query | undefined {
+    const current = this.findQuery(_class, query, options)
+    if (current !== undefined) {
+      this.removeFromQueue(current)
+      this.pushCallback(current, callback)
+
+      return current
+    }
+  }
+
+  private createQuery<T extends Doc>(
     _class: Ref<Class<T>>,
     query: DocumentQuery<T>,
     callback: (result: FindResult<T>) => void,
     options?: FindOptions<T>
-  ): () => void {
+  ): Query {
+    const queries = this.queries.get(_class) ?? []
     const result = this.client.findAll(_class, query, options)
     const q: Query = {
       _class,
       query,
       result,
       total: 0,
       options: options as FindOptions<Doc>,
-      callback: callback as (result: Doc[]) => void
+      callbacks: [callback as (result: Doc[]) => void]
     }
-    this.queries.push(q)
+    queries.push(q)
     result
-      .then((result) => {
+      .then(async (result) => {
         q.total = result.total
-        q.callback(result)
+        await this.callback(q)
       })
       .catch((err) => {
         console.log('failed to update Live Query: ', err)
       })
 
+    this.queries.set(_class, queries)
+    if (this.queue.length > CACHE_SIZE) {
+      this.remove()
+    }
+    return q
+  }
+
+  private remove (): void {
+    const q = this.queue.shift()
+    if (q === undefined) return
+    if (q.callbacks.length > 0) {
+      return this.remove()
+    }
+    const queries = this.queries.get(q._class)
```

Comment:
I supposed queue is for queries without callback, why you need code below?

Reply:
Avoid possible concurrency